### ISSUE TITLE
Data Inconsistency Issue after cutover

### ISSUE THREAD

**Issue by mayanksamrat** [github/gh-ost#420]
Hi All, 
    We just found an issue with our data consistency after doing cutover, when ghost copied 100% of the data to new table.

Here is the Command we have used for this migration- 

```
./gh-ost --user="gh_user" --password="xxxxx"  --allow-on-master --database="xxxxxx" --table="xxxxxxx" --verbose -max-lag-millis=50000 --alter="engine=innodb;"  --throttle-control-replicas=pawslmktxxxxxx:3306,pawslmktxxxxxxx:3306 --initially-drop-ghost-table --initially-drop-old-table --max-load=Threads_running=30 --chunk-size=50000 --concurrent-rowcount --switch-to-rbr --postpone-cut-over-flag-file=/tmp/ghostcutover.file --execute
```

After cutover we found the in new table less numbers are inserted. Is there any case these inserts can be ignored while using gh-ost. 

After copying all the data, we found that there is a mismatch in the number of rows of both the tables, original one and the ghost one. Here is the table counts.

```
mysql> select count(1) from sales_data;
+----------+
| count(1) |
+----------+
| 99426470 |
+----------+
1 row in set (0.00 sec)

mysql> select count(1) from _sales_data_gho;
+----------+
| count(1) |
+----------+
| 87669870 |
+----------+
1 row in set (0.00 sec)

mysql> select min(order_item_id) from sales_data;
+--------------------+
| min(order_item_id) |
+--------------------+
|          500000002 |
+--------------------+
1 row in set (0.00 sec)

mysql> select min(order_item_id) from _sales_data_gho;
+--------------------+
| min(order_item_id) |
+--------------------+
|          500000002 |
+--------------------+
1 row in set (0.00 sec)

mysql> select max(order_item_id) from sales_data;
+--------------------+
| max(order_item_id) |
+--------------------+
|         3456045067 |
+--------------------+
1 row in set (0.00 sec)

mysql> select max(order_item_id) from _sales_data_gho;
+--------------------+
| max(order_item_id) |
+--------------------+
|         3456045067 |
+--------------------+
1 row in set (0.00 sec)
```
Unfortunately we have missed the ghost screen output. 

-Mayank Tyagi 
+919971915405

**Comment by shlomi-noach** [comment]
I see you've been running this on the master. Does that mean data is still being written even as `gh-ost` completes the migration?

I'm also unsure I follow on the output here; you say the numbers are different _after_ the cut-over, but you show output from `_sales_data_gho`. The `_gho` table is the one `gh-ost` generates _during_ the migration and is renamed into the original table _at_ the cut-over. So after the cut-over this table shouldn't exist. Can you please explain exactly what was executed and when?

FYI, `gh-ost` let's you test-on-replica, see https://github.com/github/gh-ost/blob/master/doc/testing-on-replica.md
as a way to build your trust and confirm the integrity of your data is in tact. This can't be performed on a master because data keeps changing; but `gh-ost` can do that on a replica by stopping replication at end of migration.

**Comment by mayanksamrat** [comment]
Hi Shlomi, 
   Thanks for your quick response. 
  We have a two hour delay slave in our cluster, the above output is from that slave. We have stopped replication on delay slave just to identify the issue. 
   Cut over is still pending on this slave. I will try migrating the same in test-on-replica mode and will share the results with you.

-Mayank Tyagi

**Comment by shlomi-noach** [comment]
@mayanksamrat thank you for clarifying. The `--allow-on-master` setting confused me; I recommend you remove it.

So you've been running this on a replica that was stopped? `gh-ost` will issue a `start slave` for you 😛 

`test-on-replica` is your best friend here.

Also, it would be useful to dump the table structure. In particular, are you using the `JSON` or `POINT` types from `5.7`?

**Comment by mayanksamrat** [comment]
@shlomi-noach  The replica, from I have shared the out put was not stopped( I just stopped it now to avoid cut over on this machine), Its a delay slave (with 2 hour delay).  Attaching Orchestrator screen shot for the reference.

![screen shot 2017-05-18 at 11 29 09 am](https://cloud.githubusercontent.com/assets/25004816/26188549/68a5d5c6-3bbd-11e7-8731-0d23d940551f.png)

Please find the table structure below.
```
CREATE TABLE `sales_data` (
  `order_item_id` int(12) unsigned NOT NULL,
  `order_id` int(12) unsigned NOT NULL,
  `product_id` int(12) unsigned NOT NULL,
  `merchant_id` int(12) unsigned NOT NULL,
  `item_status` tinyint(4) unsigned NOT NULL,
  `qty_ordered` smallint(5) unsigned NOT NULL,
  `fulfillment_id` int(12) unsigned DEFAULT NULL,
  `shipper_id` int(12) unsigned DEFAULT NULL,
  `promo_code` varchar(255) DEFAULT NULL,
  `mrp` int(12) unsigned DEFAULT NULL,
  `price` int(12) unsigned DEFAULT '0',
  `discount` int(12) unsigned DEFAULT '0',
  `selling_price` int(12) unsigned DEFAULT '0',
  `max_refund` int(12) unsigned DEFAULT '0',
  `isLMD` tinyint(1) DEFAULT NULL,
  `isCOD` tinyint(1) DEFAULT NULL,
  `wallet_paid` int(12) unsigned DEFAULT '0',
  `cod_paid` int(12) unsigned DEFAULT '0',
  `pg_paid` int(12) unsigned DEFAULT '0',
  `created_at` timestamp NOT NULL DEFAULT '0000-00-00 00:00:00',
  `updated_at` timestamp NOT NULL DEFAULT '0000-00-00 00:00:00',
  `payout_status` tinyint(2) DEFAULT '1',
  `fulfillment_created_at` datetime DEFAULT NULL,
  `custom_text_1` varchar(255) DEFAULT NULL,
  `fulfillment_response` varchar(255) DEFAULT NULL,
  `tracking_number` varchar(255) DEFAULT NULL,
  `vertical_id` int(12) unsigned DEFAULT NULL,
  `refunded` tinyint(1) DEFAULT '0',
  `reverse_payout_id` int(11) DEFAULT NULL,
  `shipping_amount` int(12) unsigned DEFAULT '0',
  `sku_name` varchar(255) DEFAULT NULL,
  `conv_fee` int(12) unsigned DEFAULT '0',
  `delivered_at` datetime DEFAULT NULL,
  `adjustments` int(12) DEFAULT '0',
  `shipped_at` datetime DEFAULT NULL,
  `customer_id` int(12) unsigned DEFAULT NULL,
  `customer_type` tinyint(4) unsigned DEFAULT NULL,
  `custom_text_2` varchar(255) DEFAULT NULL,
  `create_timestamp` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
  `update_timestamp` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  `ship_by_date` datetime DEFAULT NULL,
  `fulfillment_service` int(12) unsigned DEFAULT '0',
  `payout_status_forward` enum('SYNCED','REVENUE_CREATED','REVENUE_FAILED','REVENUE_AUTO_CLOSE','PAYOUT_CREATED','PAYOUT_PAID','PAYOUT_HOLD','PAYOUT_CARRY_FORWARD','PAYOUT_IN_PROCESS') DEFAULT 'SYNCED',
  `payout_status_reverse` enum('SYNCED','REVENUE_CREATED','REVENUE_FAILED','REVENUE_AUTO_CLOSE','PAYOUT_CREATED','PAYOUT_PAID','PAYOUT_HOLD','PAYOUT_CARRY_FORWARD','PAYOUT_IN_PROCESS') DEFAULT 'SYNCED',
  PRIMARY KEY (`order_item_id`),
  KEY `updated_at` (`updated_at`),
  KEY `created_at` (`created_at`),
  KEY `merchant_id` (`merchant_id`),
  KEY `order_id` (`order_id`),
  KEY `fulfillment_created_at` (`fulfillment_created_at`),
  KEY `reverse_payout_id` (`reverse_payout_id`),
  KEY `fulfillment_service` (`fulfillment_service`),
  KEY `tracking_number` (`tracking_number`),
  KEY `idx_product_id` (`product_id`),
  KEY `idx_update_timestamp` (`update_timestamp`),
  KEY `delivered_at` (`delivered_at`),
  KEY `shipped_at` (`shipped_at`),
  KEY `idx_sd_payout_status_forward` (`payout_status_forward`),
  KEY `idx_sd_payout_status_reverse` (`payout_status_reverse`),
  KEY `customer_id_idx` (`customer_id`),
  KEY `customer_type_idx` (`customer_type`),
  KEY `mer_created_idx` (`merchant_id`,`created_at`),
  KEY `idx_sd_ship_by_date` (`ship_by_date`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8
```

-Mayank Tyagi

**Comment by shlomi-noach** [comment]
Bonus +2 points for using `orchestrator`!

OK, very good; everything still makes sense, since the replicas, even delayed, may still have applied changes in the duration of the `cut-over` and so your test was not atomic.

**Comment by mayanksamrat** [comment]
@shlomi-noach, Cut-over has been done now on delay machine too,  Still we have a mismatch data for sales_data and _sales_data_del table, the same mismatch we have found on every database server of the cluster.

 The old table has more rows than the new table that ghost has created. Check the sample logs below.

```
mysql> select count(1) from _sales_data_del where order_item_id > 3431616560 and order_item_id < 3431766933;
+----------+
| count(1) |
+----------+
|     8671 |
+----------+
1 row in set (0.00 sec)

mysql> select count(1) from sales_data where order_item_id > 3431616560 and order_item_id < 3431766933;
+----------+
| count(1) |
+----------+
|        0 |
+----------+
1 row in set (0.00 sec)

```
Seems It has missed some batches to be inserted into the migrated table. 

-Mayank Tyagi

**Comment by shlomi-noach** [comment]
@mayanksamrat, I'm sorry, this still doesn't make sense.

Did you now run `--test-on-replica`? Because with `--test-on-replica` there _is no_ `_del` table. With `--test-on-replica` you are left with a `_gho` table. I suspect you were not using `--test-on-replica` and as such, again, the table is still changing; there is no point to the comparison.
I think we're going in circles around this one. Please check again the docs on https://github.com/github/gh-ost/blob/master/doc/testing-on-replica.md

Also, how about those replication filters? What are they filtering?

Also, you're sharing output from a delayed replica. Why is that? Can you please share the data from a normal replica please? Let's eliminate extra factors.

**Comment by shlomi-noach** [comment]
Also, to clarify, with `--test-on-replica` you will compare the tables _on the replica you connected to_ -- the data will be unavailable on any other replica.

**Comment by mayanksamrat** [comment]
@shlomi-noach  Let me clear here. 

Last day I tried to run ghost using --allow on master, there is a delay slave in that cluster. after cut over on master, we found the data mismatch in the _del and migrated table. 
Now I have shared the logs from delay slave only. But after cutover done on delay slave too, there was a mismatch(Shared the table counts with you already). We have revert back the changes by renaming the _del table to the original one and applied all the transactions happened after cutover.


Now we are running test-on-replica to reproduce the same if we can found something.

-Mayank Tyagi

**Comment by shlomi-noach** [comment]
Thank you for clarifying!

Can you please elaborate on the replication filters?

**Comment by mayanksamrat** [comment]
Only MySQL database is not being replicated here.

**Comment by mayanksamrat** [comment]
Hi @shlomi-noach, 
      We run the test-on-replica on one of our slaves. It worked fine this time. Counts and ckecksum for the tables (original and ghost one) are same. Still not getting why data mismatch happened in allow-on-master mode.

```
mysql> select count(1) from _xxxx_xxxx_gho;
+----------+
| count(1) |
+----------+
| 99961572 |
+----------+
1 row in set (46.44 sec)

mysql> select count(1) from xxxx_xxxx;
+----------+
| count(1) |
+----------+
| 99961572 |
+----------+
1 row in set (53.78 sec)
```

-Mayank Tyagi

**Comment by shlomi-noach** [comment]
@mayanksamrat the `--test-on-replica` runs the exact same algorithm as the master migration, with a minor diff in the duration of the `cut-over` phase (though both use the same `cut-over` algorithm).

I'd be surprised to hear there's a difference in behavior between `--test-on-replica` and actual master migration.

Please continue to run `--test-on-replica` a few times, to either gain confidence that all is well or to reproduce the problem you had.
